### PR TITLE
A major (performance) update on the submodule: srl-zoo

### PR DESCRIPTION
**Pull request by ncble**
### Image rotated by 90 degrees
- see [issue](https://github.com/araffin/robotics-rl-srl/issues/41)

### Image preprocessing
Instead of using specific image preprocessing (mode='imagenet') as 

        x[..., 0] -= 0.485
        x[..., 1] -= 0.456
        x[..., 2] -= 0.406
        # Scaling
        x[..., 0] /= 0.229
        x[..., 1] /= 0.224
        x[..., 2] /= 0.225

the mode='tf' that rescale pixel value to [-1,1] is preferable and could avoid unwanted effect when applying training with the other datasets/networks.

### CODE REVIEW

**Comment by ncble** [comment]
Since the [srl_zoo](https://github.com/araffin/srl-zoo) is less popular than robotic-rl-srl, I decided to post the changelog [here](https://github.com/araffin/robotics-rl-srl/pull/50):

## Highlights

1. **2~5 times speed-up (overall)** (srl_zoo) compared to the current version of origin/master
2. Better SRL training mechanism (more intuitive, **better modularity**) that support sophisticated update (e.g. GAN)
3. Add GAN to srl_zoo
4. Scalable SRL models (support any image shape, see the detail below)
5. **~9 times faster** DataLoader which is also simpler, since it's natively supported by pytorch.
6. Remove several redundant codes to speed-up training.
7. Fix several issues of (robotics-rl_srl): #41, #42, #43, #46, #47, #48, #49...

Note: **Before, we need to modify 8 scripts in order to add one new model, now only two scripts (at most three): models/modules.py and models/my_custom_model.py (see the template models/new_model_template.py)**

## Changelog

### SRL part
- Support any image resolution for the entire toolbox (**--img-shape="(3,128,128)"**), including the DataLoader, Environment, SRL models. Before, the SRL models are not scalable with respect to image shape, and it's not sufficient to modify only the input shape (e.g. need to manually calculate each layers' shape, size, etc). Now, all models function more like keras.
- Support adversarial state representation learning. (e.g. GAN)
- [**New scripts**] models/base_trainer.py, models/new_model_template.py, models/gan.py
	- base_trainer.py: new trainig pipeline for better modularity.
	- new_model_template.py: a simple example of "how to add new model to srl_zoo". 
	- gan.py: adversarial state representation learning.

- Better (simpler, ~10 times faster) plots
- Support new monitor mode (**--monitor**) "loss" (before, there is only "pbar" progressbar): monitor losses during training, calculate GTC per epoch.
- Support control of number of CPU for dataloader (**--num-worker**).
- Support "anytime training": load the previous trained SRL model weights to continue the training **--srl-pre-weights** (weights path)
- Change validation mechanism to the classic one (i.e. within one epoch: train then valid). Before, we alternate between training and validation mode at batch level.
- Support specific GPU number. (by **--gpu-num=0**, --gpu_num1, etc)
- [**Remove**]: preprocessing/preprocess.py (it's useless), models/custom_layers.py
- [**Rename**] the models/models.py is renamed to models/base_models.py, since it's more intuitive for the outsider. Currently, there are several confusing names "custom_layers.py", "modules.py", "learner.py" "models.py 
### RL part

- support any image shape.
- support specific GPU number. (by **--gpu-num=0**, --gpu_num1, etc)
- support **--srl-model-path** indicate the SRL model weights path. Before, we can only load either the latest (by calling --latest) or manually change the config/srl_model.yaml model weights path.
- register new srl models
- fix issues:
	- image rotated by 90 degrees
	- **--log-folder** folder doesn't exist.

- [**New**] replay/plot_pipeline.py: aggregate all losses and plot on one figure. (draft code to be refined, merged with replay/aggregate_plots.py, compare_plots.py, gather_results.py)

File: losses/losses.py [hunk 1]
```diff
@@ -374,3 +379,30 @@ def tripletLoss(states, p_states, n_states, weight, loss_manager, alpha=0.2):
     tcn_triplet_loss = tcn_triplet_loss.mean()
     loss_manager.addToLosses('triplet_loss', weight, tcn_triplet_loss)
     return weight * tcn_triplet_loss
+
+
```
Comment (kalifou) [inline losses/losses.py:404]: Please add docstrings for the loss methods in the following lines.

**Comment by araffin** [comment]
Btw, why did you use a branch on your repo where you could create a branch on the original repo? (you are part of the collaborators), same for srl-robotics repo

File: losses/losses.py (hunk 1 above)
Comment (kalifou) [reply losses/losses.py:404]: Also, please reformat the file (to conform with PEP8 style).

File: models/autoencoders.py [hunk 2]
```diff
@@ -1,23 +1,37 @@
 from __future__ import print_function, division, absolute_import
```
Comment (kalifou) [inline models/autoencoders.py:1]: There is a lot of code commented out...
In some cases it appears to display what is done internaly, in other cases forgotten...Maybe needs some cleaning.

File: models/autoencoders.py [hunk 3]
```diff
@@ -114,5 +139,199 @@ def decode(self, x):
...
+    def reconstruct(self, x):
+        return self.model.decode(self.model.encode(x))
+    def encode(self, x):
+        return self.model.encode(x)
+    def decode(self, x):
+        return self.model.decode(x)
+    def forward(self, x):
+        return self.model.encode(x) ## or self.model(x)
+
+if __name__ == "__main__":
```
Comment (kalifou) [inline models/autoencoders.py:217]: Maybe this example can now be removed ?

File: models/base_models.py [hunk 4]
```diff
@@ -0,0 +1,507 @@
+from __future__ import print_function, division, absolute_import
+
```
Comment (kalifou) [inline models/base_models.py:2]: This file needs a little reformat

File: models/base_models.py [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+            crop1 = bridge
+        else:
+            crop1 = self.center_crop(bridge, up.shape[2:])
+        # print(up.shape)
+        # print(crop1.shape)
+        out = torch.cat([up, crop1], 1)
+        out = self.conv_block(out)
+
+        return out
+if __name__ == "__main__":
```
Comment (kalifou) [inline models/base_models.py:508]: maybe this example could be removed

File: models/gan.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+try:
+    # relative import
+    from .base_models import BaseModelSRL, ConvSN2d, ConvTransposeSN2d, LinearSN, UNet
+    from .base_trainer import BaseTrainer
+    from ..losses.losses import ganNonSaturateLoss, autoEncoderLoss, ganBCEaccuracy, AEboundLoss
+except:
+    from models.base_models import BaseModelSRL, ConvSN2d, ConvTransposeSN2d, LinearSN, UNet
+    from models.base_trainer import BaseTrainer
+    from losses.losses import ganNonSaturateLoss, autoEncoderLoss, ganBCEaccuracy, AEboundLoss
+
```
Comment (kalifou) [inline models/gan.py]: Please update, missing docstring here and there. Also commented code to be removed.
I would suggest in case of specific architecture to include references when possible.

File: models/learner.py [hunk 7]
```diff
@@ -172,33 +187,46 @@ def __init__(self, state_dim, model_type="resnet", inverse_model_type="linear",
...
+            if self.episode_prior:
+                learnable_params += [p for p in self.prior_discriminator.parameters()]
+            self.optimizer = torch.optim.Adam(learnable_params, lr=learning_rate)
+        else:
+            assert not self.episode_prior, "NotImplementedError"
+            self.optimizer_D = torch.optim.Adam(
+                self.module.model.discriminator.parameters(), lr=learning_rate_gan[0], betas=(0.5, 0.9))
+            self.optimizer_G = torch.optim.Adam(
+                self.module.model.generator.parameters(), lr=learning_rate_gan[1], betas=(0.5, 0.9))
+            combined_E_params  = [param for param in self.module.model.encoder.parameters()]
```
Comment (kalifou) [inline models/learner.py]: Maybe you could use a loop over here ? (param collection)

File: models/learner.py [hunk 8]
```diff
@@ -247,333 +275,465 @@ def loadSavedModel(log_folder, valid_models, cuda=True):
...
-
-        data_loader = DataLoader(minibatchlist, images_path, n_workers=N_WORKERS, multi_view=self.multi_view,
-                                 use_triplets=self.use_triplets, is_training=True, apply_occlusion=self.use_dae,
-                                 occlusion_percentage=self.occlusion_percentage)
-        test_data_loader = DataLoader(test_minibatchlist, images_path, n_workers=N_WORKERS, multi_view=self.multi_view,
-                                      use_triplets=self.use_triplets, max_queue_len=1, is_training=False,
-                                      apply_occlusion=self.use_dae, occlusion_percentage=self.occlusion_percentage)
+        print("Train: {} minibatches, {} samples".format(len(dataloader_train), len(train_set)))
+        print("Valid: {} minibatches, {} samples".format(len(dataloader_valid), len(valid_set)))
+        # =======================================================================
```
Comment (kalifou) [inline models/learner.py:448]: Construction site...I hope that you plan to make your loader complatible with the following section :)

File: models/learner.py [hunk 9]
```diff
@@ -247,333 +275,465 @@ def loadSavedModel(log_folder, valid_models, cuda=True):
...
-            loss_history['train_loss'].append(train_loss)
-            loss_history['val_loss'].append(val_loss)
-            for key in loss_history.keys():
-                if key in ['train_loss', 'val_loss']:
-                    continue
-                loss_history[key][-1] /= epoch_batches
-                if epoch + 1 < N_EPOCHS:
-                    loss_history[key].append(0)
-
+            def update_loss_history(loss_manager, train_loss, val_loss, epoch_batches, epoch):
```
Comment (kalifou) [inline models/learner.py:798]: This function should be defined somewhere else, for instance in util/ losses.py file.

File: models/learner.py [hunk 10]
```diff
@@ -247,333 +275,465 @@ def loadSavedModel(log_folder, valid_models, cuda=True):
...
+                    # if self.use_dae:
+                    #     noisy_obs = noisy_obs.to(self.device)
+                    #     next_noisy_obs = next_noisy_obs.to(self.device)    
+                        
+                    if self.model_type != "gan":
+                        ## It's extremely important to release loss tensor, otherwise it will cause 'memory leak".
+                        self.optimizer.zero_grad()
+                        loss_manager.resetLosses() 
+
+                    # Predict states given observations as in Time Contrastive Network (Triplet Loss) [Sermanet et al.]
```
Comment (kalifou) [inline models/learner.py]: Here as well, it seems that some loses/ models are not taken into account yet...

File: models/modules.py [hunk 11]
```diff
@@ -1,25 +1,29 @@
-from .autoencoders import CNNAutoEncoder, DenseAutoEncoder, LinearAutoEncoder
-from .vae import CNNVAE, DenseVAE
-from .forward_inverse import BaseForwardModel, BaseInverseModel, BaseRewardModel
+from .autoencoders import AutoEncoderTrainer
+from .vae import VAETrainer # CNNVAE, DenseVAE
+from .forward_inverse import BaseForwardModel, BaseInverseModel, BaseRewardModel, BasicTrainer
 from .priors import SRLConvolutionalNetwork, SRLDenseNetwork, SRLLinear
 from .triplet import EmbeddingNet
-from .models import *
+from .gan import GANTrainer # Generator, Discriminator, Encoder, UNet,
```
Comment (kalifou) [inline models/modules.py]: If I remember well from our discussion, this is under development...

File: models/new_model_template.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
+import torch
+import torch.nn as nn
```
Comment (kalifou) [inline models/new_model_template.py:2]: Very good idea :+1: , it should even appear in the doc if not there yet !

**Review by kalifou — COMMENTED** [review]
Hi, here is a preliminary review (on few of the modified files).
Nice to see the suggested code refactoring,
even tough I wait to see more since the code is still under development.
Please take time to also update/pass the tests in both repositories.
I will update my review soon.

File: models/learner.py [hunk 13]
```diff
@@ -44,35 +50,37 @@ class BaseLearner(object):
...
+        self.module = None
         self.seed = seed
         self.use_dae = False
+        assert not self.use_dae, "Not implemented error."
         # Seed the random generator
         np.random.seed(seed)
-        th.manual_seed(seed)
-        if cuda:
+        torch.manual_seed(seed)
+        if cuda >= 0:  # This will increase the training time by 5-10%.
```
Comment (kalifou) [inline models/learner.py:69]: Maybe it helps speed up in the case of a multi-gpu setting ? 
Not clear, misleading...

Comment (ncble) [reply models/learner.py:69]: I wanted to say, by making cuda deterministic would **decrease** the performance, and **increase** the training time by 5-10%. These lines (lines 67-70) can be comment-out if the user want further speed-up.

File: losses/losses.py (hunk 1 above)
Comment (ncble) [reply losses/losses.py:404]: Done. I used vscode autopep8.

File: models/autoencoders.py (hunk 2 above)
Comment (ncble) [reply models/autoencoders.py:1]: Done. I have cleaned the code.

File: models/autoencoders.py (hunk 3 above)
Comment (ncble) [reply models/autoencoders.py:217]: This could help when debugging, what do you think if we keep them ? (No side effect)

**Comment by ncble** [comment]
The original version (master) of srl_zoo has 5719 lines of code. My pull-request has already added and removed +4145/-1453 lines of code, thus 5598 lines of code have been modified (~97% of code). There is no need to be merged to the master branch.
